ec2/tests/test_ec2_instance.py

```diff
@@ -67,6 +73,130 @@ def mock_ctx(self, test_name):
+                               private_key_path=private_key_path)
+        self.assertIn("server can't both have a",
+                      ex.message)
+
+    @mock_ec2
+    def test_no_key_pair_file(self, *_):
+        """tests correct error is thrown when key file is missing
+        """
+
+        name = 'test_no_key_pair_file'
+        ctx = self.mock_ctx(name)
+        current_ctx.set(ctx=ctx)
+
+        ctx.node.properties['use_password'] = True
+        private_key_dir = tempfile.mkdtemp()
+        private_key_path = '{0}/{1}.pem' \
+            .format(private_key_dir, name)
+
+        with mock.patch('ec2.utils.get_single_connected_node_by_type') \
+                as mock_get_single_connected_node_by_type:
+            mock_get_single_connected_node_by_type.return_value = None
+            ex = self.assertRaises(NonRecoverableError,
+                                   instance._get_private_key,
+                                   private_key_path=private_key_path)
+            self.assertIn("Cannot find private key file; expected file path",
+                          ex.message)
+
+    @mock_ec2
+    def test_relationship_key_pair(self, *_):
+        """tests finding a key file via relationship
+        """
+
+        name = 'test_relationship_key_pair'
+        ctx = self.mock_ctx(name)
+        current_ctx.set(ctx=ctx)
+
+        ctx.node.properties['use_password'] = True
+        private_key_dir = tempfile.mkdtemp()
+        private_key_path = '{0}/{1}.pem' \
+            .format(private_key_dir, name)
+        open(private_key_path, 'w').close()
+
+        key_pair_node = MockNodeContext(properties={
+            'private_key_path': private_key_path
+        })
+
+        with mock.patch('ec2.utils.get_single_connected_node_by_type') \
+                as mock_get_single_connected_node_by_type:
+            mock_get_single_connected_node_by_type.return_value = key_pair_node
+            output = instance._get_private_key('')
+            self.assertEqual(private_key_path, output)
+
+    @mock_ec2
+    @mock.patch('ec2.utils.get_single_connected_node_by_type')
+    def test_bootstrap_context_key_pair(self, *_):
+        """tests finding a key file via bootstrap context
+        """
+
+        name = 'test_bootstrap_context_key_pair'
+        private_key_dir = tempfile.mkdtemp()
+        private_key_path = '{0}/{1}.pem' \
+            .format(private_key_dir, name)
+        open(private_key_path, 'w').close()
+
+        cloudify_agent = MockContext()
+        cloudify_agent['agent_key_path'] = private_key_path
+
+        bootstrap_ctx = BootstrapContext({
+            'cloudify_agent': cloudify_agent
+        })
+
+        ctx = MockCloudifyContext(bootstrap_context=bootstrap_ctx,
+                                  node_name=name)
+        current_ctx.set(ctx=ctx)
+
+        with mock.patch('ec2.utils.get_single_connected_node_by_type') \
+                as mock_get_connected:
+            mock_get_connected.return_value = None
```

Comment:
FYI: you can do the return_value thing in the "patch" call, e.g. - 
"with mock.patch('ec2...', return_value=None) as mock_get_connected:
...